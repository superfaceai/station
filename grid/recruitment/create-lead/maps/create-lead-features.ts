/* eslint-disable jest/no-export */
import { RecordingProcessOptions, SuperfaceTest } from '@superfaceai/testing';

import { buildSuperfaceTest } from '../../../test-config';

export const createLeadFeaturesTest = (
  provider: string,
  options?: RecordingProcessOptions
): void => {
  describe(`recruitment/create-lead/${provider}`, () => {
    describe('CreateLeadFeatures', () => {
      let superface: SuperfaceTest;

      beforeAll(() => {
        superface = buildSuperfaceTest({
          profile: 'recruitment/create-lead',
          provider,
          useCase: 'CreateLeadFeatures',
        });
      });

      it(`returns valid ${provider} features`, async () => {
        const result = await superface.run(
          {
            input: {},
          },
          options
        );

        expect(() => result.unwrap()).not.toThrow();
        expect(result).toMatchSnapshot();
      });
    });
  });
};
