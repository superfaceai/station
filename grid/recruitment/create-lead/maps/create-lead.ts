/* eslint-disable jest/no-export, jest/valid-describe-callback, jest/valid-title, jest/no-identical-title */

import { BinaryData, IMappedError } from '@superfaceai/one-sdk';
import { RecordingProcessOptions, SuperfaceTest } from '@superfaceai/testing';

import { buildSuperfaceTest } from '../../../test-config';

const sampleLead = {
  name: 'John Lead',
  firstName: 'John',
  lastName: 'Lead',

  email: 'john.lead@fakemail.com',
  phone: '1-859-557-6573',
  address: '25772 Gustave Shore, Iowa, USA',

  education: [
    {
      degree: 'MBA',
      school: 'University of Pennsylvania',
      fieldOfStudy: null,
      startedAt: '2008-03-01',
      endedAt: '2011-03-30',
    },
    {
      degree: 'B.S.',
      school: 'University of Chicago',
      fieldOfStudy: 'Marketing Communication & Economics',
      startedAt: '2004-09-01',
      endedAt: '2007-03-30',
    },
  ],

  workExperience: [
    {
      position: 'Sales Director',
      summary: 'Summary of a work experience at Test Company',
      company: 'Test Company',
      industry: 'Telecommunications',
      current: false,
      startedAt: '2011-03-01',
      endedAt: '2014-03-30',
    },
  ],

  cv: {
    fileName: 'cv-sample.pdf',
    mimeType: 'application/pdf',
    data: BinaryData.fromPath('./grid/recruitment/cv-sample.pdf'),
  },

  links: [
    {
      name: 'twitter',
      url: 'https://twitter.com/candidate.username',
    },
    {
      name: 'Portfolio',
      url: 'https://url.to.portfolio',
    },
  ],

  source: ['LinkedIn', 'Indeed', 'Glassdoor'],
};

type Error = {
  title: string;
  detail?: unknown;
  code: string;
  rateLimit?: {
    bucket?: string;
    totalRequests?: number;
    remainingRequests?: number;
    remainingRequestsPercentage?: number;
    resetTimestam?: number;
  };
};

export const createLeadTest = (
  provider: string,
  jobIds: { valid: string; invalid: string },
  options?: RecordingProcessOptions
): void => {
  describe(`recruitment/create-lead/${provider}`, () => {
    let superface: SuperfaceTest;
    const JEST_TIMEOUT = 20 * 1000;

    beforeEach(() => {
      superface = buildSuperfaceTest({
        profile: 'recruitment/create-lead',
        provider,
      });
    });

    describe('CreateLead', () => {
      describe('when specified job does exist', () => {
        it(
          'should perform successfully',
          async () => {
            const result = await superface.run(
              {
                useCase: 'CreateLead',
                input: {
                  jobId: jobIds.valid,
                  ...sampleLead,
                },
              },
              options
            );

            expect(() => result.unwrap()).not.toThrow();
            expect(result).toMatchSnapshot();
          },
          JEST_TIMEOUT
        );
      });

      describe('when specified job does not exist', () => {
        it('should map error', async () => {
          const result = await superface.run(
            {
              useCase: 'CreateLead',
              input: {
                jobId: jobIds.invalid,
                firstName: 'Demo',
                lastName: 'Testing',
                email: 'demo_testing@fakemail.com',
              },
            },
            {
              fullError: true,
              ...options,
            }
          );

          expect(() => result.unwrap()).toThrow();
          result.match(
            () => {},
            err => {
              expect(
                (err as unknown as IMappedError<Error>).properties?.code
              ).toBe('JobNotFound');
            }
          );
        });
      });
    });
  });
};

export const createLeadBreezyHRSpecificTest = (
  options?: RecordingProcessOptions
): void => {
  const provider = 'breezy-hr';

  describe(`recruitment/create-lead/${provider}`, () => {
    let superface: SuperfaceTest;

    describe('CreateLead', () => {
      beforeAll(() => {
        superface = buildSuperfaceTest({
          profile: 'recruitment/create-lead',
          provider,
          useCase: 'CreateLead',
        });
      });

      describe('when specified company does not exist', () => {
        let companyId: string | undefined;

        beforeAll(() => {
          companyId = process.env.BREEZY_HR_COMPANY_ID;

          process.env.BREEZY_HR_COMPANY_ID = '1b111c1111ef11';
        });

        afterAll(() => {
          process.env.BREEZY_HR_COMPANY_ID = companyId;
        });

        it('should map error', async () => {
          const result = await superface.run(
            {
              useCase: 'CreateLead',
              input: {
                jobId: 'JOB_ID',
                firstName: 'Demo',
                lastName: 'Testing',
                email: 'demo_testing@fakemail.com',
              },
            },
            {
              fullError: true,
              ...options,
            }
          );

          expect(() => result.unwrap()).toThrow();
          result.match(
            () => {},
            err => {
              expect((err as IMappedError<Error>).properties?.code).toBe(
                'WrongIntegrationParameter'
              );
            }
          );
        });
      });
    });
  });
};

export const createLeadWorkableSpecificTest = (
  options?: RecordingProcessOptions
): void => {
  const provider = 'workable';

  describe(`recruitment/create-lead/${provider}`, () => {
    let superface: SuperfaceTest;

    describe('CreateLead', () => {
      beforeAll(() => {
        superface = buildSuperfaceTest({
          profile: 'recruitment/create-lead',
          provider,
          useCase: 'CreateLead',
        });
      });

      describe('when specified subdomain does not exist', () => {
        let subdomain: string | undefined;

        beforeAll(() => {
          subdomain = process.env.WORKABLE_SUBDOMAIN;

          process.env.WORKABLE_SUBDOMAIN = 'invalid-subdomain';
        });

        afterAll(() => {
          process.env.WORKABLE_SUBDOMAIN = subdomain;
        });

        it('returns error', async () => {
          const result = await superface.run(
            {
              useCase: 'CreateLead',
              input: {
                jobId: 'JOB_ID',
                firstName: 'Demo',
                lastName: 'Testing',
                email: 'demo_testing@fakemail.com',
              },
            },
            {
              fullError: true,
              ...options,
            }
          );

          expect(() => result.unwrap()).toThrow();
          result.match(
            () => {},
            err => {
              expect((err as IMappedError<Error>).properties?.code).toBe(
                'WrongIntegrationParameter'
              );
            }
          );
        });
      });
    });
  });
};
