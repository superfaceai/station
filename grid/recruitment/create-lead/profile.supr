"""
Create Lead
Create sourced lead in ATS (Applicant Tracking System).
"""
name = "recruitment/create-lead"
version = "2.0.0"

"""
Create Lead
Creates a sourced prospect candidate either for some specific job or any job within a company.
"""
usecase CreateLead {

  input {
    jobId
    firstName!
    lastName!
    email!
    phone
    address 
    education 
    workExperience
    links
    cv
    source
  }

  result {
    id!
    jobId
    rateLimit
  }!

  error Error

  example Successful {
    input {
      jobId = 'JOB_ID'
      firstName = 'Demo'
      lastName = 'Testing'
      email = 'demo_testing@fakemail.com'
    }

    result {
      id = 'LEAD_ID'
      jobId = 'JOB_ID'
    }
  }

  example Failed {
    input {
      jobId = 'NOT_EXISTING_JOB_ID'
      firstName = 'Demo'
      lastName = 'Testing'
      email = 'demo_testing@fakemail.com'
    }

    error {
      title = 'Bad Request'
      detail = 'Job with id NOT_EXISTING_JOB_ID does not exist.'
      code = 'JobNotFound'
      rateLimit = {
        bucket = 'workable'
      }
    }
  }
}

"""
Get Create Lead Features
Get features of CreateLead use case supported by the provider.
"""
usecase CreateLeadFeatures {
  result {
    """
    CV MIME Types
    List of CV MIME types accepted by the provider. MIME types are defined in RFC2045 and registered by https://www.iana.org/.
    """
    cvMIMETypes! [string!]!
    """
    CV Upload Methods
    List of supported CV upload methods.
    """
    cvUploadMethods! [
      enum {
        """
        URL
        The cv input contains url field with publicly accessible URL of a CV file.
        """
        url = 'url'
        """
        File
        The cv input contains data field with CV file and fileName field with the file name.
        """
        file = 'file'
      }!
    ]!
 }

  example Successful {
    result {
      cvMIMETypes = ['application/pdf', 'text/rtf']
      cvUploadMethods = ['url', 'file']
    }
  }
}

"""
Id
Unique identifier of a lead
"""
field id string!

"""
Job Id
Unique identifier of a job that the lead is a good prospect for.
"""
field jobId string!

"""
Job Title
Title of a job, lead might apply for.
"""
field jobTitle string!

"""
First Name
First name of a lead.
"""
field firstName string!

"""
Last Name
Last name of a lead.
"""
field lastName string!

"""
Full Name
Full name of a lead.
"""
field fullName string

"""
Email
Email of a lead.
"""
field email string!

"""
Phone
Phone number of a lead.
"""
field phone string

"""
Address
Address of a lead.
"""
field address string

"""
Education
List of sourced education experiences relevant to lead.
"""
field education [Education]

"""
Work Experience
List of sourced work experiences relevant to lead.
"""
field workExperience [WorkExperience]

"""
Links
List of sourced links relevant to lead. Can contain links to social profiles, portfolios or other.
"""
field links [Link]

"""
CV (Resume)
Object containing either URL to a CV or data with CV document.
"""
field cv {
  """
  URL
  Publicly accessible URL of a CV file.
  """
  url string 
  """
  File Name
  The file name of the CV document.
  """
  fileName string
  """
  MIME Type
  MIME type of the CV document.
  """
  mimeType string
  """
  Data
  Binary content of the CV document.
  """
  data
}

"""
Source
Name or list of names, with sources from where this lead comes from.
"""
field source string | [string]

"""
Created At
ISO8601 timestamp when was a lead created.
"""
field createdAt string

"""
Updated At
ISO8601 timestamp when was a lead lastly updated.
"""
field updatedAt string

"""
Rate limit
Rate limit details.
"""
field rateLimit RateLimit

"""
Education
Object representing educational experience of a lead.
"""
model Education {
  """
  Id
  Unique identifier of lead's education experience.
  """
  id string
  
  """
  School
  Name of a school.
  """
  school string

  """
  Degree
  Degree aquired by finishing lead's education.
  """
  degree string

  """
  Field of Study
  Name of a field studied in lead's education.
  """
  fieldOfStudy string
  
  """
  Started At
  ISO8601 timestamp when the education started.
  """
  startedAt string

  """
  Ended At
  ISO8601 timestamp when the education ended.
  """
  endedAt string
}

"""
Work Experience
Object representing work experience of a lead.
"""
model WorkExperience {
  """
  Id
  Unique identifier of lead's work experience.
  """
  id string

  """
  Position
  Name of position, lead has occupied.
  """
  position string

  """
  Summary
  Summary of lead's work experience. 
  """
  summary string

  """
  Company
  Name of company.
  """
  company string

  """
  Industry
  Name of industry, the company worked in.
  """
  industry string

  """
  Current
  Boolean representing whether lead still work there.
  """
  current boolean

  """
  Started At
  ISO8601 timestamp when lead started working.
  """
  startedAt string

  """
  Ended At
  ISO8601 timestamp when lead stoped working.
  """
  endedAt string
}

"""
Link
Object representing a link sourced with lead.
"""
model Link {
  """
  Name
  Descriptive name of a link to some URL.
  """
  name string

  """
  URL
  Actual URL destination of the link.
  """
  url string
}

model Error {
  """
  Error Title
  A short, human-readable summary of the problem type.
  """
  title!
  """
  Error Detail
  A human-readable explanation specific to this occurrence of the problem.
  """
  detail
  """
  Code
  Error code.
  """
  code! ErrorCode!
  
  rateLimit 
}

model RateLimit {
  """
  Bucket
  Different parts of API may have different rate limits.
  Bucket identifies to which part of API the rate limits apply.
  """
  bucket string
  """
  Total requests
  Total requests available in the time window.
  """
  totalRequests number
  """
  Remaining requests
  Remaining requests available in the time window.
  """
  remainingRequests number
  """
  Remaining requests available in percents
  Remaining requests available in the time window in percents.
  """
  remainingRequestsPercentage number
  """
  Reset timestamp
  Timestamp when the rate limits will reset (in  Unix Timestamp format). Available only for providers without rolling rate limit window.
  """
  resetTimestamp number
}

model ErrorCode enum {
  """
  Rate limit reached
  Returned when a request cannot be served due to the user or app reached max requests quota.
  """
  RateLimitReached
  """
  Unauthenticated
  There was a problem authenticating your request. Check that you are passing valid credentials.
  """
  Unauthenticated
  """
  Unauthorized
  The request is understood, but it has been refused or access is not allowed.
  """
  Unauthorized
  """
  Unknown error
  Returned when an unexpected error occured.
  """
  UnknownError
  """
  CV file name required
  Missing cv.fileName input. CV file name is required when cv.data input field is defined.
  """
  CVFileNameRequired
  """
  CV MIME type not supported
  CV MIME type is not supported by the provider. To get supported MIME types run CreateLeadFeatures use case.
  """
  CVMIMETypeNotSupported
  """
  Wrong integration parameter
  The integration parameter value has not been accepted.
  """
  WrongIntegrationParameter
  """
  Job not found
  The job ID does not exist.
  """
  JobNotFound
}
