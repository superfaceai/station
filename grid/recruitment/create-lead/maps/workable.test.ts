import { createLeadTest, createLeadWorkableSpecificTest } from './create-lead';
import { createLeadFeaturesTest } from './create-lead-features';

const workableProviderName = 'workable';

createLeadTest(workableProviderName, {
  valid: 'EEF7355636',
  invalid: 'NOT-EXISTING',
});

createLeadWorkableSpecificTest();

createLeadFeaturesTest(workableProviderName);
