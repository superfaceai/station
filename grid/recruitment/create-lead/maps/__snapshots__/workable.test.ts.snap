// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recruitment/create-lead/workable CreateLead when specified job does exist should perform successfully 1`] = `
Ok {
  "value": Object {
    "id": "dbe7f28",
    "jobId": "EEF7355636",
    "rateLimit": Object {
      "bucket": "workable",
      "remainingRequests": 8,
      "remainingRequestsPercentage": 80,
      "resetTimestamp": 1675094240,
      "totalRequests": 10,
    },
  },
}
`;

exports[`recruitment/create-lead/workable CreateLeadFeatures returns valid workable features 1`] = `
Ok {
  "value": Object {
    "cvMIMETypes": Array [
      "application/pdf",
      "text/rtf",
      "application/msword",
      "application/vnd.openxmlformats-officedocument.custom-properties+xml",
    ],
    "cvUploadMethods": Array [
      "url",
      "file",
    ],
  },
}
`;
